1.20.5-rc1 - adding modded ore crashes the game

Since updating from 1.20.5-pre1 to pre3, I have been getting crashes on my three mods that add ore gen to the game.

https://gist.github.com/kwpugh/87739bee296c3d30188706eab6c0fe9f

Same occurs in pre4 and now rc1

in chatting with TelepathicGrunt, it was pointed out the modded features might not be getting registered by the api.

![Screenshot 2024-04-18 at 9 59 06 PM](https://github.com/FabricMC/fabric/assets/8809210/1b978624-6566-4ace-8ca1-d200007abe2b)

My ore gen setup is pretty standard, following the approach from KaupenJoe's ore gen tutorial

```
[18:27:35] [Server thread/ERROR] (Minecraft) Error executing task on Chunk source main thread executor for minecraft:overworld
 net.minecraft.ReportedException: Biome decoration
	at net.minecraft.server.level.ChunkMap.method_39934(ChunkMap.java:691) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:162) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.level.ServerChunkCache$MainThreadExecutor.doRunTask(ServerChunkCache.java:561) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:136) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.level.ServerChunkCache$MainThreadExecutor.pollTask(ServerChunkCache.java:570) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.level.ServerChunkCache.pollTask(ServerChunkCache.java:278) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:829) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:817) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:145) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:792) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:696) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.MinecraftServer.method_29739(MinecraftServer.java:275) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 93
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100) ~[?:?]
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106) ~[?:?]
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302) ~[?:?]
	at java.base/java.util.Objects.checkIndex(Objects.java:385) ~[?:?]
	at java.base/java.util.ArrayList.get(ArrayList.java:427) ~[?:?]
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[?:?]
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102) ~[minecraft-merged-2052ebfe04-1.20.5-rc2-loom.mappings.1_20_5_rc2.layered+hash.2198-v2.jar:?]
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808) ~[?:?]
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188) ~[?:?]
---- Minecraft Crash Report ----
// Surprise! Haha. Well, this is awkward.

Time: 2024-04-19 18:27:35
Description: Biome decoration

java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 93
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356)
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100)
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148)
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62)
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91)
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146)
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Worker-Main-16
Stacktrace:
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356)
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100)
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148)

-- Generation --
Details:
	CenterX: 2
	CenterZ: -37
	Decoration Seed: 7583121984018732223
Stacktrace:
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356)
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100)
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148)
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62)
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91)
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146)
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

-- Chunk to be generated --
Details:
	Status being generated: minecraft:features
	Location: 2,-37
	Position hash: -158913789950
	Generator: net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator@6ee0f1b0
Stacktrace:
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62)
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91)
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146)
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

-- System Details --
Details:
	Minecraft Version: 1.20.5 Release Candidate 2
	Minecraft Version ID: 1.20.5-rc2
	Operating System: Windows 11 (amd64) version 10.0
	Java Version: 21.0.2, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode, sharing), Oracle Corporation
	Memory: 880988744 bytes (840 MiB) / 2818572288 bytes (2688 MiB) up to 8573157376 bytes (8176 MiB)
	CPUs: 24
	Processor Vendor: AuthenticAMD
	Processor Name: AMD Ryzen 9 5900X 12-Core Processor            
	Identifier: AuthenticAMD Family 25 Model 33 Stepping 2
	Microarchitecture: Zen 3
	Frequency (GHz): 3.69
	Number of physical packages: 1
	Number of physical CPUs: 12
	Number of logical CPUs: 24
	Graphics card #0 name: NVIDIA GeForce RTX 2060 SUPER
	Graphics card #0 vendor: NVIDIA
	Graphics card #0 VRAM (MB): 8192.00
	Graphics card #0 deviceId: VideoController1
	Graphics card #0 versionInfo: 31.0.15.5123
	Memory slot #0 capacity (MB): 16384.00
	Memory slot #0 clockSpeed (GHz): 3.20
	Memory slot #0 type: DDR4
	Memory slot #1 capacity (MB): 16384.00
	Memory slot #1 clockSpeed (GHz): 3.20
	Memory slot #1 type: DDR4
	Virtual memory max (MB): 65457.04
	Virtual memory used (MB): 24624.55
	Swap memory total (MB): 32768.00
	Swap memory used (MB): 193.13
	JVM Flags: 0 total; 
	Fabric Mods: 
		exotelcraft: Exotelcraft 6.0.0-ALPHA-1
		fabric-api: Fabric API 0.97.3+1.20.5
		fabric-api-base: Fabric API Base 0.4.40+80f8cf51d1
		fabric-api-lookup-api-v1: Fabric API Lookup API (v1) 1.6.59+e9d2a72bd1
		fabric-biome-api-v1: Fabric Biome API (v1) 13.0.23+b21c00cbd1
		fabric-block-api-v1: Fabric Block API (v1) 1.0.20+6dfe4c9bd1
		fabric-block-view-api-v2: Fabric BlockView API (v2) 1.0.8+80f8cf51d1
		fabric-blockrenderlayer-v1: Fabric BlockRenderLayer Registration (v1) 1.1.50+80f8cf51d1
		fabric-client-tags-api-v1: Fabric Client Tags 1.1.12+7f945d5bd1
		fabric-command-api-v1: Fabric Command API (v1) 1.2.45+f71b366fd1
		fabric-command-api-v2: Fabric Command API (v2) 2.2.24+80f8cf51d1
		fabric-commands-v0: Fabric Commands (v0) 0.2.62+df3654b3d1
		fabric-content-registries-v0: Fabric Content Registries (v0) 8.0.2+c5fc38b3d1
		fabric-convention-tags-v1: Fabric Convention Tags 2.0.1+7f945d5bd1
		fabric-convention-tags-v2: Fabric Convention Tags (v2) 1.0.0+7f945d5bd1
		fabric-crash-report-info-v1: Fabric Crash Report Info (v1) 0.2.27+80f8cf51d1
		fabric-data-attachment-api-v1: Fabric Data Attachment API (v1) 1.1.14+2a2c66b6d1
		fabric-data-generation-api-v1: Fabric Data Generation API (v1) 19.0.3+7f945d5bd1
		fabric-dimensions-v1: Fabric Dimensions API (v1) 2.1.68+94793913d1
		fabric-entity-events-v1: Fabric Entity Events (v1) 1.6.8+e9d2a72bd1
		fabric-events-interaction-v0: Fabric Events Interaction (v0) 0.7.6+c5fc38b3d1
		fabric-game-rule-api-v1: Fabric Game Rule API (v1) 1.0.50+80f8cf51d1
		fabric-gametest-api-v1: Fabric Game Test API (v1) 1.3.15+1172e897d1
		fabric-item-api-v1: Fabric Item API (v1) 8.1.1+17e985d6d1
		fabric-item-group-api-v1: Fabric Item Group API (v1) 4.0.37+aae0949ad1
		fabric-key-binding-api-v1: Fabric Key Binding API (v1) 1.0.45+80f8cf51d1
		fabric-keybindings-v0: Fabric Key Bindings (v0) 0.2.43+df3654b3d1
		fabric-lifecycle-events-v1: Fabric Lifecycle Events (v1) 2.3.4+c5fc38b3d1
		fabric-loot-api-v2: Fabric Loot API (v2) 3.0.3+97f703dad1
		fabric-message-api-v1: Fabric Message API (v1) 6.0.10+109a837cd1
		fabric-model-loading-api-v1: Fabric Model Loading API (v1) 1.0.12+80f8cf51d1
		fabric-models-v0: Fabric Models (v0) 0.4.11+9386d8a7d1
		fabric-networking-api-v1: Fabric Networking API (v1) 4.0.7+9342ba64d1
		fabric-object-builder-api-v1: Fabric Object Builder API (v1) 15.1.2+c5fc38b3d1
		fabric-particles-v1: Fabric Particles (v1) 4.0.0+c5fc38b3d1
		fabric-recipe-api-v1: Fabric Recipe API (v1) 5.0.2+c5fc38b3d1
		fabric-registry-sync-v0: Fabric Registry Sync (v0) 5.0.13+aae0949ad1
		fabric-renderer-api-v1: Fabric Renderer API (v1) 3.2.12+97f703dad1
		fabric-renderer-indigo: Fabric Renderer - Indigo 1.5.12+80f8cf51d1
		fabric-renderer-registries-v1: Fabric Renderer Registries (v1) 3.2.61+df3654b3d1
		fabric-rendering-data-attachment-v1: Fabric Rendering Data Attachment (v1) 0.3.46+73761d2ed1
		fabric-rendering-fluids-v1: Fabric Rendering Fluids (v1) 3.1.3+2c869dedd1
		fabric-rendering-v0: Fabric Rendering (v0) 1.1.64+df3654b3d1
		fabric-rendering-v1: Fabric Rendering (v1) 4.2.4+b21c00cbd1
		fabric-resource-conditions-api-v1: Fabric Resource Conditions API (v1) 4.0.1+74e2f560d1
		fabric-resource-loader-v0: Fabric Resource Loader (v0) 1.0.4+18de49c6d1
		fabric-screen-api-v1: Fabric Screen API (v1) 2.0.21+7b70ea8ad1
		fabric-screen-handler-api-v1: Fabric Screen Handler API (v1) 1.3.70+b21c00cbd1
		fabric-sound-api-v1: Fabric Sound API (v1) 1.0.21+c5fc38b3d1
		fabric-transfer-api-v1: Fabric Transfer API (v1) 5.1.6+c5fc38b3d1
		fabric-transitive-access-wideners-v1: Fabric Transitive Access Wideners (v1) 6.0.10+74e2f560d1
		fabricloader: Fabric Loader 0.15.10
		java: Java HotSpot(TM) 64-Bit Server VM 21
		minecraft: Minecraft 1.20.5-rc.2
		mixinextras: MixinExtras 0.3.5
---- Minecraft Crash Report ----
// I feel sad now :(

Time: 2024-04-19 18:27:36
Description: Biome decoration

java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 93
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356)
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100)
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148)
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62)
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91)
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146)
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)


A detailed walkthrough of the error, its code path and all known details is as follows:
---------------------------------------------------------------------------------------

-- Head --
Thread: Worker-Main-5
Stacktrace:
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:100)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:106)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:302)
	at java.base/java.util.Objects.checkIndex(Objects.java:385)
	at java.base/java.util.ArrayList.get(ArrayList.java:427)
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356)
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100)
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148)

-- Generation --
Details:
	CenterX: 2
	CenterZ: -37
	Decoration Seed: 7583121984018732223
Stacktrace:
	at net.minecraft.world.level.chunk.ChunkGenerator.applyBiomeDecoration(ChunkGenerator.java:356)
	at net.minecraft.world.level.chunk.status.ChunkStatusTasks.generateFeatures(ChunkStatusTasks.java:100)
	at net.minecraft.world.level.chunk.status.ChunkStatus.generate(ChunkStatus.java:148)
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62)
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91)
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146)
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

-- Chunk to be generated --
Details:
	Status being generated: minecraft:features
	Location: 2,-37
	Position hash: -158913789950
	Generator: net.minecraft.world.level.levelgen.NoiseBasedChunkGenerator@6ee0f1b0
Stacktrace:
	at net.minecraft.server.level.ChunkMap.method_17224(ChunkMap.java:675)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482)
	at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.method_17634(ChunkTaskPriorityQueueSorter.java:62)
	at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:91)
	at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:146)
	at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:102)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)

-- System Details --
Details:
	Minecraft Version: 1.20.5 Release Candidate 2
	Minecraft Version ID: 1.20.5-rc2
	Operating System: Windows 11 (amd64) version 10.0
	Java Version: 21.0.2, Oracle Corporation
	Java VM Version: Java HotSpot(TM) 64-Bit Server VM (mixed mode, sharing), Oracle Corporation
	Memory: 880988744 bytes (840 MiB) / 2818572288 bytes (2688 MiB) up to 8573157376 bytes (8176 MiB)
	CPUs: 24
	Processor Vendor: AuthenticAMD
	Processor Name: AMD Ryzen 9 5900X 12-Core Processor            
	Identifier: AuthenticAMD Family 25 Model 33 Stepping 2
	Microarchitecture: Zen 3
	Frequency (GHz): 3.69
	Number of physical packages: 1
	Number of physical CPUs: 12
	Number of logical CPUs: 24
	Graphics card #0 name: NVIDIA GeForce RTX 2060 SUPER
	Graphics card #0 vendor: NVIDIA
	Graphics card #0 VRAM (MB): 8192.00
	Graphics card #0 deviceId: VideoController1
	Graphics card #0 versionInfo: 31.0.15.5123
	Memory slot #0 capacity (MB): 16384.00
	Memory slot #0 clockSpeed (GHz): 3.20
	Memory slot #0 type: DDR4
	Memory slot #1 capacity (MB): 16384.00
	Memory slot #1 clockSpeed (GHz): 3.20
	Memory slot #1 type: DDR4
	Virtual memory max (MB): 65457.04
	Virtual memory used (MB): 24624.55
	Swap memory total (MB): 32768.00
	Swap memory used (MB): 193.13
	JVM Flags: 0 total; 
	Fabric Mods: 
		exotelcraft: Exotelcraft 6.0.0-ALPHA-1
		fabric-api: Fabric API 0.97.3+1.20.5
		fabric-api-base: Fabric API Base 0.4.40+80f8cf51d1
		fabric-api-lookup-api-v1: Fabric API Lookup API (v1) 1.6.59+e9d2a72bd1
		fabric-biome-api-v1: Fabric Biome API (v1) 13.0.23+b21c00cbd1
		fabric-block-api-v1: Fabric Block API (v1) 1.0.20+6dfe4c9bd1
		fabric-block-view-api-v2: Fabric BlockView API (v2) 1.0.8+80f8cf51d1
		fabric-blockrenderlayer-v1: Fabric BlockRenderLayer Registration (v1) 1.1.50+80f8cf51d1
		fabric-client-tags-api-v1: Fabric Client Tags 1.1.12+7f945d5bd1
		fabric-command-api-v1: Fabric Command API (v1) 1.2.45+f71b366fd1
		fabric-command-api-v2: Fabric Command API (v2) 2.2.24+80f8cf51d1
		fabric-commands-v0: Fabric Commands (v0) 0.2.62+df3654b3d1
		fabric-content-registries-v0: Fabric Content Registries (v0) 8.0.2+c5fc38b3d1
		fabric-convention-tags-v1: Fabric Convention Tags 2.0.1+7f945d5bd1
		fabric-convention-tags-v2: Fabric Convention Tags (v2) 1.0.0+7f945d5bd1
		fabric-crash-report-info-v1: Fabric Crash Report Info (v1) 0.2.27+80f8cf51d1
		fabric-data-attachment-api-v1: Fabric Data Attachment API (v1) 1.1.14+2a2c66b6d1
		fabric-data-generation-api-v1: Fabric Data Generation API (v1) 19.0.3+7f945d5bd1
		fabric-dimensions-v1: Fabric Dimensions API (v1) 2.1.68+94793913d1
		fabric-entity-events-v1: Fabric Entity Events (v1) 1.6.8+e9d2a72bd1
		fabric-events-interaction-v0: Fabric Events Interaction (v0) 0.7.6+c5fc38b3d1
		fabric-game-rule-api-v1: Fabric Game Rule API (v1) 1.0.50+80f8cf51d1
		fabric-gametest-api-v1: Fabric Game Test API (v1) 1.3.15+1172e897d1
		fabric-item-api-v1: Fabric Item API (v1) 8.1.1+17e985d6d1
		fabric-item-group-api-v1: Fabric Item Group API (v1) 4.0.37+aae0949ad1
		fabric-key-binding-api-v1: Fabric Key Binding API (v1) 1.0.45+80f8cf51d1
		fabric-keybindings-v0: Fabric Key Bindings (v0) 0.2.43+df3654b3d1
		fabric-lifecycle-events-v1: Fabric Lifecycle Events (v1) 2.3.4+c5fc38b3d1
		fabric-loot-api-v2: Fabric Loot API (v2) 3.0.3+97f703dad1
		fabric-message-api-v1: Fabric Message API (v1) 6.0.10+109a837cd1
		fabric-model-loading-api-v1: Fabric Model Loading API (v1) 1.0.12+80f8cf51d1
		fabric-models-v0: Fabric Models (v0) 0.4.11+9386d8a7d1
		fabric-networking-api-v1: Fabric Networking API (v1) 4.0.7+9342ba64d1
		fabric-object-builder-api-v1: Fabric Object Builder API (v1) 15.1.2+c5fc38b3d1
		fabric-particles-v1: Fabric Particles (v1) 4.0.0+c5fc38b3d1
		fabric-recipe-api-v1: Fabric Recipe API (v1) 5.0.2+c5fc38b3d1
		fabric-registry-sync-v0: Fabric Registry Sync (v0) 5.0.13+aae0949ad1
		fabric-renderer-api-v1: Fabric Renderer API (v1) 3.2.12+97f703dad1
		fabric-renderer-indigo: Fabric Renderer - Indigo 1.5.12+80f8cf51d1
		fabric-renderer-registries-v1: Fabric Renderer Registries (v1) 3.2.61+df3654b3d1
		fabric-rendering-data-attachment-v1: Fabric Rendering Data Attachment (v1) 0.3.46+73761d2ed1
		fabric-rendering-fluids-v1: Fabric Rendering Fluids (v1) 3.1.3+2c869dedd1
		fabric-rendering-v0: Fabric Rendering (v0) 1.1.64+df3654b3d1
		fabric-rendering-v1: Fabric Rendering (v1) 4.2.4+b21c00cbd1
		fabric-resource-conditions-api-v1: Fabric Resource Conditions API (v1) 4.0.1+74e2f560d1
		fabric-resource-loader-v0: Fabric Resource Loader (v0) 1.0.4+18de49c6d1
		fabric-screen-api-v1: Fabric Screen API (v1) 2.0.21+7b70ea8ad1
		fabric-screen-handler-api-v1: Fabric Screen Handler API (v1) 1.3.70+b21c00cbd1
		fabric-sound-api-v1: Fabric Sound API (v1) 1.0.21+c5fc38b3d1
		fabric-transfer-api-v1: Fabric Transfer API (v1) 5.1.6+c5fc38b3d1
		fabric-transitive-access-wideners-v1: Fabric Transitive Access Wideners (v1) 6.0.10+74e2f560d1
		fabricloader: Fabric Loader 0.15.10
		java: Java HotSpot(TM) 64-Bit Server VM 21
		minecraft: Minecraft 1.20.5-rc.2
		mixinextras: MixinExtras 0.3.5

Process finished with exit code -1073740791 (0xC0000409)
```

--- a/fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/impl/biome/modification/BiomeModificationContextImpl.java
+++ b/fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/impl/biome/modification/BiomeModificationContextImpl.java
@@ -103,6 +103,10 @@ public class BiomeModificationContextImpl implements BiomeModificationContext {
 		spawnSettings.freeze();
 	}
 
+	boolean shouldRebuildFeatures() {
+		return generationSettings.rebuildFeatures;
+	}
+
 	private class WeatherContextImpl implements WeatherContext {
 		@Override
 		public void setPrecipitation(boolean hasPrecipitation) {
@@ -194,7 +198,7 @@ public class BiomeModificationContextImpl implements BiomeModificationContext {
 		private final Registry<PlacedFeature> features = registries.get(RegistryKeys.PLACED_FEATURE);
 		private final GenerationSettings generationSettings = biome.getGenerationSettings();
 
-		private boolean rebuildFlowerFeatures;
+		boolean rebuildFeatures;
 
 		/**
 		 * Unfreeze the immutable lists found in the generation settings, and make sure they're filled up to every
@@ -204,7 +208,7 @@ public class BiomeModificationContextImpl implements BiomeModificationContext {
 			unfreezeCarvers();
 			unfreezeFeatures();
 
-			rebuildFlowerFeatures = false;
+			rebuildFeatures = false;
 		}
 
 		private void unfreezeCarvers() {
@@ -225,7 +229,7 @@ public class BiomeModificationContextImpl implements BiomeModificationContext {
 			freezeCarvers();
 			freezeFeatures();
 
-			if (rebuildFlowerFeatures) {
+			if (rebuildFeatures) {
 				rebuildFlowerFeatures();
 			}
 		}
@@ -267,7 +271,7 @@ public class BiomeModificationContextImpl implements BiomeModificationContext {
 
 			if (features.removeIf(feature -> feature.value() == placedFeature)) {
 				featureSteps.set(stepIndex, RegistryEntryList.of(features));
-				rebuildFlowerFeatures = true;
+				rebuildFeatures = true;
 
 				return true;
 			}
@@ -288,7 +292,7 @@ public class BiomeModificationContextImpl implements BiomeModificationContext {
 			featureSteps.set(index, plus(featureSteps.get(index), getEntry(features, entry)));
 
 			// Ensure the list of flower features is up-to-date
-			rebuildFlowerFeatures = true;
+			rebuildFeatures = true;
 		}
 
 		@Override

--- a/fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/impl/biome/modification/BiomeModificationImpl.java
+++ b/fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/impl/biome/modification/BiomeModificationImpl.java
@@ -26,6 +26,7 @@ import java.util.function.Consumer;
 import java.util.function.Predicate;
 
 import com.google.common.base.Stopwatch;
+import com.google.common.base.Suppliers;
 import org.jetbrains.annotations.TestOnly;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -36,6 +37,7 @@ import net.minecraft.registry.RegistryKey;
 import net.minecraft.registry.RegistryKeys;
 import net.minecraft.util.Identifier;
 import net.minecraft.world.biome.Biome;
+import net.minecraft.world.gen.feature.util.PlacedFeatureIndexer;
 
 import net.fabricmc.fabric.api.biome.v1.BiomeModificationContext;
 import net.fabricmc.fabric.api.biome.v1.BiomeSelectionContext;
@@ -152,6 +154,18 @@ public class BiomeModificationImpl {
 			// Re-freeze and apply certain cleanup actions
 			if (modificationContext != null) {
 				modificationContext.freeze();
+
+				if (modificationContext.shouldRebuildFeatures()) {
+					impl.get(RegistryKeys.DIMENSION).stream().forEach(dimensionOptions -> {
+						dimensionOptions.chunkGenerator().indexedFeaturesListSupplier = Suppliers.memoize(
+							() -> PlacedFeatureIndexer.collectIndexedFeatures(
+									List.copyOf(dimensionOptions.chunkGenerator().getBiomeSource().getBiomes()),
+									(biomeEntry) -> (biomeEntry.value().generationSettings).getFeatures(),
+									true
+							)
+						);
+					});
+				}
 			}
 		}
 

--- a/fabric-biome-api-v1/src/main/resources/fabric-biome-api-v1.accesswidener
+++ b/fabric-biome-api-v1/src/main/resources/fabric-biome-api-v1.accesswidener
@@ -55,3 +55,7 @@ accessible field net/minecraft/world/biome/GenerationSettings flowerFeatures Lja
 mutable field net/minecraft/world/biome/GenerationSettings flowerFeatures Ljava/util/function/Supplier;
 accessible field net/minecraft/world/biome/GenerationSettings allowedFeatures Ljava/util/function/Supplier;
 mutable field net/minecraft/world/biome/GenerationSettings allowedFeatures Ljava/util/function/Supplier;
+
+# ChunkGenerator
+accessible field net/minecraft/world/gen/chunk/ChunkGenerator indexedFeaturesListSupplier Ljava/util/function/Supplier;
+mutable field net/minecraft/world/gen/chunk/ChunkGenerator indexedFeaturesListSupplier Ljava/util/function/Supplier;

